src/EFCore/DbContext.cs

```diff
@@ -690,6 +690,106 @@ await DbContextDependencies.UpdateLogger.OptimisticConcurrencyExceptionAsync(thi
         /// </summary>
         public event EventHandler<SaveChangesFailedEventArgs>? SaveChangesFailed;
 
+        /// <summary>
+        ///     <para>
+        ///         An event fired when this context instance is leased from the context pool.
+        ///     </para>
+        ///     <para>
+        ///         This event is only fired when 'DbContext' pooling is enabled through use of
+        ///         <see cref="EntityFrameworkServiceCollectionExtensions.AddDbContextPool{TContext}(IServiceCollection,Action{DbContextOptionsBuilder},int)"/>
+        ///         or
+        ///         <see cref="EntityFrameworkServiceCollectionExtensions.AddPooledDbContextFactory{TContext}(IServiceCollection,Action{DbContextOptionsBuilder},int)"/>.
+        ///     </para>
+        /// </summary>
+        public event Action<DbContext>? LeasedFromPool;
+
+        /// <summary>
+        ///     Called to fire the <see cref="LeasedFromPool"/> event. Can be overriden in a derived context to intercept this event.
+        /// </summary>
+        protected virtual void OnLeasedFromPool()
+            => LeasedFromPool?.Invoke(this);
+
+        /// <summary>
+        ///     <para>
+        ///         An event fired when this context instance is leased from the context pool.
+        ///     </para>
+        ///     <para>
+        ///         This event is only fired when 'DbContext' pooling is enabled through use of
+        ///         <see cref="EntityFrameworkServiceCollectionExtensions.AddDbContextPool{TContext}(IServiceCollection,Action{DbContextOptionsBuilder},int)"/>
+        ///         or
+        ///         <see cref="EntityFrameworkServiceCollectionExtensions.AddPooledDbContextFactory{TContext}(IServiceCollection,Action{DbContextOptionsBuilder},int)"/>.
+        ///     </para>
+        /// </summary>
+        public event Func<DbContext, CancellationToken, Task>? LeasedFromPoolAsync;
+
+        /// <summary>
+        ///     Called to fire the <see cref="OnLeasedFromPoolAsync"/> event. Can be overriden in a derived context to intercept this event.
+        /// </summary>
+        /// <param name="cancellationToken"> A <see cref="CancellationToken" /> to observe while waiting for the task to complete. </param>
+        /// <returns> A task that represents the asynchronous operation. </returns>
+        /// <exception cref="OperationCanceledException"> If the <see cref="CancellationToken" /> is canceled. </exception>
+        protected virtual async Task OnLeasedFromPoolAsync(CancellationToken cancellationToken)
+        {
+            var handler = LeasedFromPoolAsync;
+            if (handler != null)
+            {
+                foreach (var func in handler.GetInvocationList())
+                {
+                    await ((Func<DbContext, CancellationToken, Task>)func)(this, cancellationToken);
+                }
+            }
```

Comment:
I'd split the part that does foreach/await into a local method, so that in the common scenario of not having any event, we don't pay the overhead of an async state machine:

```suggestion
            return LeasedFromPoolAsync is null
                ? Task.CompletedTask
                : OnLeasedFromPoolAsyncCore(cancellationToken);

            async Task OnLeasedFromPoolAsyncCore(CancellationToken cancellationToken)
            {
                var handler = LeasedFromPoolAsync;
                if (handler != null)
                {
                    foreach (var func in handler.GetInvocationList())
                    {
                        await ((Func<DbContext, CancellationToken, Task>)func)(this, cancellationToken);
                    }
                }
            }
```